dashboard/app/models/pd/session.rb

```diff
@@ -55,4 +55,33 @@ def formatted_date_with_start_and_end_times
   def hours
     (self.end - start) / 1.hour
   end
+
+  def assign_code
+    update! code: unused_random_code
+  end
+
+  def remove_code
+    update! code: nil
+  end
+
+  def self.find_by_code(code)
+    return nil unless code
+    find_by(code: code)
+  end
+
+  def open_for_attendance?
+    code.present? &&
+      workshop.state == Pd::Workshop::STATE_IN_PROGRESS &&
+      start - 24.hours < Time.zone.now &&
+      self.end + 24.hours > Time.zone.now
+  end
+
+  private
+
+  def unused_random_code
+    loop do
+      code = SectionHelpers.random_code
+      return code unless Pd::Session.exists?(code: code)
```

Comment:
It doesn't really matter if there are duplicate codes, since they're used in different contexts. I want to reuse [SectionHelpers.random_code](https://github.com/code-dot-org/code-dot-org/blob/staging/lib/cdo/section_helpers.rb#L9) because it generates a random consonant code, which is what we want here. It is odd to be called SectionHelpers here though. What if I extract that random consonant code generating logic into a more generic lib?

Also I just realized these are supposed to be 4 chars long (in the spec). I'll update.